### PR TITLE
Clang warning and compilation fixes

### PR DESCRIPTION
**Pull request by akva2**
One compilation fix (missing <array> include) and various warning quells.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams please

File: Apps/Common/SIMSolver.h [hunk 1]
```diff
@@ -155,7 +155,7 @@ template<class T1> class SIMSolver : public SIMSolverStat<T1>
     this->printHeading(heading);
 
     // Solve for each time step up to final time
-    for (int iStep = 1; this->advanceStep(); iStep++)
+    for (; this->advanceStep(); )
```
Comment (kmokstad) [inline Apps/Common/SIMSolver.h:158]: Then rather use `while (this->advanceStep())` instead

File: src/ASM/ASMutils.C [hunk 2]
```diff
@@ -176,7 +176,7 @@ bool ASM::readXML (std::istream& is, IntMat& MNPC, std::vector<Vec3>& nodes,
   {
     Vec3 X;
     std::istringstream iss(data);
-    for (size_t inod = 0; iss; inod++)
+    for ([[maybe_unused]] size_t inod = 0; iss; inod++)
```
Comment (kmokstad) [inline src/ASM/ASMutils.C:179]: This was a new thing I did not know about, but think I would prefer the (maybe old-fashioned):
```
#if SP_DEBUG > 1
  for (size_t inod = 0; iss; inod++)
#else
  while (iss)
#endif
```
However, I don't insist.

File: src/ASM/ASMutils.C [hunk 3]
```diff
@@ -194,7 +194,7 @@ bool ASM::readXML (std::istream& is, IntMat& MNPC, std::vector<Vec3>& nodes,
   {
     IntVec mnpc(nenod);
     std::istringstream iss(data);
-    for (size_t iel = 0; iss; iel++)
+    for ([[maybe_unused]] size_t iel = 0; iss; iel++)
```
Comment (kmokstad) [inline src/ASM/ASMutils.C:197]: same

File: src/ASM/BasisFunctionCache.C [hunk 4]
```diff
@@ -13,6 +13,8 @@
 
 #include "BasisFunctionCache.h"
 
+#include <array>
```
Comment (kmokstad) [inline src/ASM/BasisFunctionCache.C:16]: But `std::array` is also referred in the header file, so it makes more sense to include it there instead in my view.

File: src/Utility/HDF5Restart.C [hunk 5]
```diff
@@ -69,8 +69,13 @@ bool HDF5Restart::writeData (const SerializeData& data)
 #endif
 
   // Lambda function for writing data to the HDF5 file.
-  auto&& writeGroup = [this,pid,ptot] (hid_t group, const std::string& name,
-                                       int len, const void* data, hid_t type)
+#ifdef HAVE_MPI
+  auto&& writeGroup = [this,pid,ptot]
```
Comment (kmokstad) [inline src/Utility/HDF5Restart.C:73]: Hmm, to me it looks like `this` and `ptot` also are referred outside the `#ifdef HAVE_MPI`. Have you tested this with hdf5 support?

**Review by kmokstad — APPROVED** [review]
Mostly ok, except for a few questions

File: src/Utility/HDF5Restart.C (hunk 5 above)
Comment (akva2) [reply src/Utility/HDF5Restart.C:73]: it's not. it is used in the for immediately after the lambda definition so maybe you got confused by that?

File: Apps/Common/SIMSolver.h (hunk 1 above)
Comment (akva2) [reply Apps/Common/SIMSolver.h]: alright. you have preferred butchered for's before but i'll gladly use while :)

File: src/ASM/ASMutils.C (hunk 2 above)
Comment (akva2) [reply src/ASM/ASMutils.C]: okay.

File: src/ASM/ASMutils.C (hunk 3 above)
Comment (akva2) [reply src/ASM/ASMutils.C]: okay.

**Comment by akva2** [comment]
jenkins build this with downstreams please

File: src/Utility/HDF5Restart.C (hunk 5 above)
Comment (kmokstad) [reply src/Utility/HDF5Restart.C:73]: right, the use was only outside the lambda I can se now

**Pull request merged**